x/callback/abci.go

```diff
@@ -0,0 +1,94 @@
+package callback
+
+import (
+	"fmt"
+	"runtime/debug"
+
+	"cosmossdk.io/collections"
+	abci "github.com/cometbft/cometbft/abci/types"
+	"github.com/cometbft/cometbft/libs/log"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/archway-network/archway/x/callback/keeper"
+	"github.com/archway-network/archway/x/callback/types"
+)
+
+// EndBlocker fetches all the callbacks registered for the current block height and executes them
+func EndBlocker(ctx sdk.Context, k keeper.Keeper, wk types.WasmKeeperExpected) []abci.ValidatorUpdate {
+	logger := k.Logger(ctx)
+	params, err := k.Params.Get(ctx)
+	if err != nil {
+		panic(err)
+	}
+
+	currentHeight := ctx.BlockHeight()
+	// fetching all callbacks for current height
+	callbacks, err := k.GetCallbacksByHeight(ctx, currentHeight)
+	if err != nil {
+		panic(err)
+	}
+
+	for _, callback := range callbacks {
+		// creating CallbackMsg which is encoded to json and passed as input to contract execution
+		callbackMsg := types.NewCallbackMsg(callback.GetJobId())
+		// handling any panics
+		defer recoverAnyPanics(logger, callback)()
+		logger.Debug(
+			"executing callback",
+			"contract_address", callback.ContractAddress,
+			"job_id", callback.GetJobId(),
+			"msg", callbackMsg.String(),
+		)
+
+		// creating a child context with limited gas meter based on configured params
+		childCtx, commit := ctx.WithGasMeter(sdk.NewGasMeter(params.GetCallbackGasLimit())).CacheContext()
```

Comment:
I have one question: when the user schedules a callback, should they pay the full `params.GetCallbackGasLimit()` upfront on that tx? I'm thinking about scenarios where the `callback_gas_limit` is way higher than the actual gas the contract consumes on the callback.

Reply:
probably commit to paying the gas limit, but get refunded when not all of that is consumed

Reply:
That would be ideal. Does the SDK support returning unused gas?

Reply:
yes we can compute it